jujugui/static/gui/src/app/init/test-utils.js

```diff
@@ -99,6 +99,55 @@ describe('init utils', () => {
     });
   });
 
+  describe('jujushellURL', () => {
+    // Set up mock sources from which to retrieve the jujushell URL.
+    const setupSources = values => {
+      const config = {jujushellURL: values.config || ''};
+      const db = {
+        environment: {
+          get: sinon.stub().withArgs('jujushellURL').returns(values.db || '')
+        }
+      };
+      const storage = {
+        getItem: sinon.stub().withArgs('jujushell-url').returns(
+          values.storage || '')
+      };
+      return {
+        config: config,
+        db: db,
+        storage: storage
+      };
+    };
+
+    // Check that a call to jujushellURL in a scenario set up with the given
+    // values returns the expected URL.
+    const assertURL = (values, expectedURL) => {
+      const params = setupSources(values);
+      const url = utils.jujushellURL(params.storage, params.db, params.config);
+      assert.strictEqual(url, expectedURL);
+    };
+
+    it('returns the URL stored in the storage', () => {
+      assertURL({storage: 'wss://1.2.3.4/ws1/'}, 'wss://1.2.3.4/ws1/');
+      assertURL({storage: 'ws://1.2.3.4/ws2/'}, 'ws://1.2.3.4/ws2/');
+      assertURL({storage: '1.2.3.4/ws3/'}, 'wss://1.2.3.4/ws3/');
+      assertURL({storage: 'http://1.2.3.4/ws4/'}, 'ws://1.2.3.4/ws4/');
+      assertURL({storage: 'https://1.2.3.4/ws5/'}, 'wss://1.2.3.4/ws5/');
+    });
+
+    it('returns the URL stored in the db', () => {
+      assertURL({db: '1.2.3.4'}, 'ws://1.2.3.4/ws/');
```

Comment:
based on this test, I was expecting my value in the input box (10.0.0.1) to turn into wss://10.0.0.1/ws, however checking the failed wss connection (only shows in chrome dev tools) it tried to connect to wss://10.0.0.1/

Reply:
This is a test for when the address is obrained from the mega-watcher. In which case, we ad /ws/ as we can assume the path. In the other cases, users are free to set the path manually and we don't force /ws/.

Reply:
Let's chat about this.